src/distance.cpp

```diff
@@ -0,0 +1,620 @@
+        //If start and end are facing the same direction
+        int64_t dNoRev = abs(prefix_sum->at(i1) - prefix_sum->at(i2)); 
+        return dNoRev;
+    } else {
+        //TODO Currently assumes no reversing edges in the chain
+        return -1;
+    }
+}
+
+int64_t ChainDistances::chainDistance_short(VG* graph, pair<id_t, bool> start,
+                                                      pair<id_t, bool> end) {
+    //Distance between end of start node to beginning of end node in chain
+    
+    size_t i1 = snarl_to_index->at(start.first);
+    size_t i2 = snarl_to_index->at(end.first); 
+    int64_t d1 = chainDistance(start, end);
+    int64_t d2 = chainDistance(end, start);
+    if (d1 == -1 && d2 == -1) {
+        return -1;
+    } else if (d1 == -1) {
+        return d2 - graph->get_node(i1)->sequence().size();
+    } else if (d2 == -1) {
+        return d1 - graph->get_node(i2)->sequence().size();
+    } else {
+        return min(d2 - graph->get_node(i1)->sequence().size(), 
+                   d1 - graph->get_node(i2)->sequence().size());
+    }
+}
+
+int64_t ChainDistances::chainLength() {
+    //Get the length of a chain including length of last node
+    return prefix_sum->back();
+}
+
+
+int64_t calculateIndex(DistanceIndex* di, VG* graph, SnarlManager* sm, 
+         const Chain* chain) {
+    /*Add chain to distance index and recursively add all distances in 
+      snarls contained within chain
+    */
+    auto cmp = [] (pair<const handle_t*,int64_t> x,
+                                           pair<const handle_t*,int64_t> y) {
+        //Comparison function for the priority of a pair of handle, distance
+        return (x.second < y.second); 
+    };
+
+    vector<int64_t> chain_prefix_sum (1,0); //initialize to [0]
+
+    unordered_map<id_t, size_t> chain_snarl_to_index;
+    for (int i = 0; i < chain->size(); i++) { 
+        //for each snarl in the chain TODO ChainIterator?
+        const Snarl* snarl = chain->at(i);
+        chain_snarl_to_index.insert( make_pair<id_t, size_t>
+                     (snarl->start().node_id(), chain_snarl_to_index.size()));
+        //TODO use visit as key to identify a node+direction? or node id? 
+        //Get all nodes in snarl includes both boundary nodes of child snarls
+        pair<unordered_set<Node*>, unordered_set<Edge*>> contents = 
+                        sm->shallow_contents(snarl, *graph, true);
+        unordered_set<Node*> allNodes = contents.first;
+        /*TODO better way of getting all the nodes in the snarl? Currently
+          don't include end boundary node of a child snarl in distance 
+          matrix because I don't think I need them?
+        */ 
+
+         // Create SnarlDistance object with all nodes in this snarl
+        unordered_map <pair<id_t, bool>, size_t> vti;
+        size_t snarlIndex = 0;  
+        for (unordered_set<Node*>::iterator n = allNodes.begin(); 
+            n != allNodes.end(); ++n) {
```

Comment:
`iterator`s are a great use case for the `auto` keyword (in general it's handy for variables that have long names). You can rewrite this as 
```
for (auto n = allNodes.begin(); n != allNodes.end(); ++n) {
}
```
Or even cleaner:
```
for (const Node* node : allNodes) {
}
```